### PR TITLE
Add scan sample using subgroup operations

### PR DESCRIPTION
**Pull request by DuncanMcBain**

### CODE REVIEW

File: samples/subgroup-scan.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+  auto version = d.get_platform().get_info<sycl::info::platform::version>();
+  auto is_version = [&](std::string_view req_ver) {
+    return version.find(req_ver) != std::string::npos;
+  };
+  auto has_subgroups = d.is_host() || is_version("2.1") ||
+                       is_version("2.2") || is_version("3.0");
+  auto has_ext_subgroups = d.has_extension("cl_intel_subgroups") ||
+                           d.has_extension("cl_khr_subgroups");
+  if (!(has_subgroups || has_ext_subgroups)) {
```
Comment (Rbiessy) [inline samples/subgroup-scan.cpp:196]: The 2 booleans are not needed anymore. I think we can just check that `d.get_info<cl::sycl::info::device::max_num_sub_groups>() > 0` now.

**Review by Rbiessy — APPROVED** [review]

File: samples/subgroup-scan.cpp (hunk 1 above)
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp:196]: This does not seem to work on the Intel GPU driver that I have on my laptop (though that is rather old, maybe I just need to update). Version 20.23.16988 for reference

Comment (Rbiessy) [reply samples/subgroup-scan.cpp:196]: Ok, that's annoying. I may need to change that in SYCL-DNN.

File: samples/subgroup-scan.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  std::vector<int32_t> sum(in.size());
+  {
+    sycl::buffer<int32_t, 1> buf(sycl::range<1>(in.size()));
+    buf.set_final_data(sum.data());
+    q.submit([&](sycl::handler& cgh) {
+      auto acc = buf.get_access<sycl::access::mode::write>(cgh);
+      cgh.copy(in.data(), acc);
+    });
+
+    par_scan<int32_t, sycl::experimental::plus<int32_t>>(buf, q);
```
Comment (alycm) [inline samples/subgroup-scan.cpp:161]: Experimental?

File: samples/CMakeLists.txt [hunk 3]
```diff
@@ -58,6 +58,10 @@ add_sample(TARGET placeholder-accessors SOURCES placeholder-accessors.cpp)
 add_sample(TARGET reduction SOURCES reduction.cpp)
 add_sample(TARGET reinterpret SOURCES reinterpret.cpp)
 add_sample(TARGET scan SOURCES scan.cpp)
+if (COMPUTECPP_SDK_USE_SUBGROUPS)
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (alycm) [inline samples/CMakeLists.txt:62]: I don't suppose it's really a problem, but this won't be scoped by the `if`, right?  So files before this won't be C++17 and ones after will be, it's a bit arbitrary.

File: samples/subgroup-scan.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+ *
+ *  scan-subgroups.cpp
+ *
+ *  Description:
+ *    Example of a parallel inclusive scan in SYCL using subgroup operations.
+ *
+ **************************************************************************/
+
+#include <CL/sycl.hpp>
+namespace sycl = cl::sycl;
```
Comment (hjabird) [inline samples/subgroup-scan.cpp:26]:
```suggestion
#include <sycl/sycl.hpp>
```
The `namespace sycl` bit isn't needed in 2020.

File: samples/subgroup-scan.cpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+/* Performs an inclusive scan with the given associative binary operation `Op`
+ * on the data in the `in` buffer. Runs in parallel on the provided accelerated
+ * hardware queue. Modifies the input buffer to contain the results of the scan.
+ * Input size has to be a power of two. If the size isn't so, the input can
+ * easily be padded to the nearest power of two with any values, and the scan on
+ * the meaningful part of the data will stay the same. */
+template <typename T, typename Op>
+void par_scan(sycl::buffer<T, 1> in, sycl::queue q) {
+  if ((in.get_count() & (in.get_count() - 1)) != 0 || in.get_count() == 0) {
```
Comment (hjabird) [inline samples/subgroup-scan.cpp:45]: `get_count()` is deprecated. Use `size()` instead. https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_buffer_interface table 39.

File: samples/subgroup-scan.cpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+          size_t gid = item.get_global_linear_id();
+
+          auto sub_group = item.get_sub_group();
+          auto scan_res = inclusive_scan_over_group(sub_group, data[gid], Op{});
+          if (sub_group.get_local_id() == sub_group.get_local_range() - 1) {
+            temp[sub_group.get_group_id()[0]] = scan_res;
+          }
+          item.barrier(sycl::access::fence_space::local_space);
+          for (auto i = 1u; i < sub_group.get_group_range()[0]; i++) {
+            scan_res += sub_group.get_group_id()[0] >= i ? temp[i - 1] : 0;
```
Comment (hjabird) [inline samples/subgroup-scan.cpp:98]:
```suggestion
          for (auto i = 1u; i < sub_group.get_group_linear_range(); i++) {
            scan_res += sub_group.get_group_linear_id() >= i ? temp[i - 1] : 0;
```
Same goes for most spaces where you have `get_id/range()[0]`.

File: samples/subgroup-scan.cpp (hunk 4 above)
Comment (hjabird) [reply samples/subgroup-scan.cpp:26]: I'm a bit confused - it looks as if you're not compiling/writing this as SYCL-2020, but sub-groups is a SYCL-2020 feature, and not a 1.2.1 feature?

File: samples/CMakeLists.txt (hunk 3 above)
Comment (DuncanMcBain) [reply samples/CMakeLists.txt:62]: #270 has this fix

File: samples/subgroup-scan.cpp (hunk 6 above)
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp]: Applied locally

File: samples/subgroup-scan.cpp [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+int main() {
+  sycl::queue q{sycl::default_selector{}};
+  sycl::device d = q.get_device();
+
+  auto version = d.get_platform().get_info<sycl::info::platform::version>();
+  auto is_version = [&](std::string_view req_ver) {
+    return version.find(req_ver) != std::string::npos;
+  };
+  auto has_subgroups = d.is_host() || is_version("2.1") ||
+                       is_version("2.2") || is_version("3.0");
```
[the remark this answers is not in the document]
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp]: Removed as part of 2020 change

File: samples/subgroup-scan.cpp [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  auto has_ext_subgroups = d.has_extension("cl_intel_subgroups") ||
+                           d.has_extension("cl_khr_subgroups");
+  if (!(has_subgroups || has_ext_subgroups)) {
+    std::cout << "Subgroups not supported on device.\n";
+    return 0;
+  }
+
+  auto ret = test_sum(q);
+  if (ret != 0) {
+    return ret;
```
[the remark this answers is not in the document]
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp:194]: Added "wrong answer" output

File: samples/subgroup-scan.cpp (hunk 5 above)
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp:44]: Has anyone made a bitesize joke yet?

File: samples/subgroup-scan.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
+/***************************************************************************
+ *
+ *  Copyright (C) 2017 Codeplay Software Limited
```
[the remark this answers is not in the document]
Comment (alycm) [reply samples/subgroup-scan.cpp]: Blank is fine.

File: samples/CMakeLists.txt (hunk 3 above)
Comment (alycm) [reply samples/CMakeLists.txt:62]: Cool, thanks.

File: samples/subgroup-scan.cpp (hunk 2 above)
Comment (alycm) [reply samples/subgroup-scan.cpp]: Ah, I don't know!  I assume the updated version works though.

File: samples/subgroup-scan.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (alycm) [reply samples/subgroup-scan.cpp:194]: Thanks.

File: samples/subgroup-scan.cpp [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  return 0;
+}
+
+int main() {
+  sycl::queue q{sycl::default_selector{}};
+  sycl::device d = q.get_device();
+
+  auto version = d.get_platform().get_info<sycl::info::platform::version>();
```
[the remark this answers is not in the document]
Comment (alycm) [reply samples/subgroup-scan.cpp]: That works too.

File: samples/subgroup-scan.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+void par_scan(sycl::buffer<T, 1> in, sycl::queue q) {
+  if ((in.get_count() & (in.get_count() - 1)) != 0 || in.get_count() == 0) {
+    throw std::runtime_error("Given input size is not a power of two.");
+  }
+
+  auto dev = q.get_device();
+
+  // Check if there is enough global memory.
+  size_t global_mem_size = dev.get_info<sycl::info::device::global_mem_size>();
+  if (!dev.is_host() && in.get_count() > (global_mem_size)) {
```
[the remark this answers is not in the document]
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp]: I should clarify: I did remove it :smile:

File: samples/subgroup-scan.cpp (hunk 2 above)
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp]: It does indeed. @hjabird should I prefer `experimental::` or just `sycl::`?

Comment (hjabird) [reply samples/subgroup-scan.cpp]: `sycl`.

File: samples/subgroup-scan.cpp [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+  if (wgroup_size == 0) {
+    throw std::runtime_error(
+        "Could not find an appropriate work-group size for the given input.");
+  }
+
+  using namespace sycl::access;
+  q.submit([&](sycl::handler& cgh) {
+    auto data = in.template get_access<mode::read_write>(cgh);
+    auto temp = sycl::accessor<T, 1, mode::read_write, target::local>(
+        sycl::range<1>(wgroup_size), cgh);
```
Comment (hjabird) [inline samples/subgroup-scan.cpp:85]: You might also consider making this a `local_accessor<T,1>`. `target::local` accessors are deprecated.

**Review by hjabird — APPROVED** [review]
Looks good!

File: samples/subgroup-scan.cpp (hunk 2 above)
Comment (DuncanMcBain) [reply samples/subgroup-scan.cpp]: Thanks, I'll leave it is as now then!

**Pull request merged**